### ISSUE TITLE
Suggest When Unused Generics Exist

### ISSUE THREAD

**Issue by JasonBock** [dotnet/roslyn#44330]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0 (Latest)

#### Diagnostic ID

No diagnostic ID

#### Repro steps

1. Create a new .NET Core 3.1 console app
2. Create a class in the project like this:
```
public class UnusedGeneric<T>
{
   public int DoSomething<TVal>(int a) => a * 2;
}
```

#### Expected behavior
I'd like to see some kind of error/warning/suggestion that `T` and `TVal` are not being used.

#### Actual behavior
No indication is given.

Would be nice if it was similar to IDE0060 when you have unused method parameters.

**Comment by mavasani** [comment]
This might also be a good candidate to be an IDE code style analyzer similar to our existing remove unnecessary analyzers (unused using, unused private members, unused parameters, unused value assignment, unreachable dead code, etc.).

**Comment by mavasani** [comment]
Tagging @sharwell

**Comment by CyrusNajmabadi** [comment]
Note: generics are sometimes used in static contexts as holders.  i.e. `Telemetry<T> { public static SomeMetricValue { get; } }`.  I don't use T here, but it allows me to associate data with a specific type in a very clean and simple fashion.

**Comment by JasonBock** [comment]
@CyrusNajmabadi that's a case that I didn't consider. I admit, I'm not sure why this is a good thing, but apparently this is a technique that is used. So, the type and its usage would be this?
```
public static class Telemetry<T>
{
  public static SomeMetric Value { get; set; }
}

...

Telemetry<CustomerService>.Value = new SomeMetric(...);
```
This almost feels like a dictionary approach, where the key is the type, and the value is the metric. Is that correct?

That said (and this is biased to my limited view), would the majority of cases show up as parameters that are no longer used? IOW, a refactoring occurred where the generic was used, but now is unused, and can be removed (similar to an unused parameter)

**Comment by mavasani** [comment]
IMO, we should still fire diagnostics/suggestions for such cases, and a source suppression with a proper justification comment would act as a good documentation on the explicit intent of use of an unused type parameter for holders. We can also consider adding an option to allow users to turn off these suggestions for static generic types.

**Comment by CyrusNajmabadi** [comment]
> This almost feels like a dictionary approach, where the key is the type, and the value is the metric. Is that correct?

Yup!  That's correct. A value you get out of this is that it's very fast and threadsafe.

**Comment by Youssef1313** [comment]
I see the comments are discussing this as an analyzer. However, the issue is marked as a warning wave.

Will this be implemented in the IDE-side or Compilers-side?

@CyrusNajmabadi @mavasani @gafter

**Comment by CyrusNajmabadi** [comment]
> Will this be implemented in the IDE-side or Compilers-side?

There design have even been established. As I pointed out in my posts, a type parameter may not be used explicitly, but may still be viral w to the correct behavior of the component.
